app/src/main/java/com/artemzin/qualitymatters/network/HostSelectionInterceptor.java

```diff
@@ -0,0 +1,54 @@
+package com.artemzin.qualitymatters.network;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import okhttp3.HttpUrl;
+import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
+
+/**
+ * Such implementation allows us easily change base url in the integration and functional tests!
+ */
+public class HostSelectionInterceptor implements Interceptor {
+
+    @Nullable
+    private volatile String baseUrl;
+
+    @Override @NonNull
+    public Response intercept(@NonNull Chain chain) throws IOException {
+        Request request = chain.request();
+        final String baseUrl = this.baseUrl;
+
+        if (baseUrl != null) {
+            request = modifyRequest(request, baseUrl);
```

Comment:
I prefer to recreate `Retrofit` than `OkHttp`, because `baseUrl` is something depends on `Retrofit` not `OkHttp`. If we change `baseUrl` in an `Intercepter`, we add magic to our code. It's not really good.

Actually, what I'm doing in my project is: Using some kind of TestModule so that we can inject mock or fake url for testing.